core/parentchain/light-client/src/io.rs

```diff
@@ -51,48 +57,69 @@ impl<B: Block, Client: Decode + Encode + Debug> StaticSealedIO for LightClientSe
-				return init_validator::<B>(params)
-			}
-
-			let validator =
-				LightClientSeal::<B, GrandpaLightValidation<B>>::unseal_from_static_file()?;
-
-			let genesis = validator.genesis_hash(validator.num_relays()).unwrap();
-			if genesis == params.get_genesis_header().hash() {
-				info!("Found already initialized light client with Genesis Hash: {:?}", genesis);
-				info!("light client state: {:?}", validator);
-				Ok(validator.latest_finalized_header(validator.num_relays()).unwrap())
-			} else {
-				init_validator::<B>(params)
-			}
-		},
+	if SgxFile::open(LIGHT_CLIENT_DB).is_err() {
+		info!("[Enclave] ChainRelay DB not found, creating new! {}", LIGHT_CLIENT_DB);
+		return init_validator::<B, OCallApi>(params, ocall_api)
+	}
+
+	let validation_state =
+		LightClientStateSeal::<B, LightValidationState<B>>::unseal_from_static_file()?;
+
+	let relay = validation_state
+		.tracked_relays
+		.get(&validation_state.num_relays)
+		.ok_or(Error::NoSuchRelayExists)?;
+
+	let genesis = relay.header_hashes[0];
+
+	if genesis == params.get_genesis_header().hash() {
+		let mut validator = init_validator::<B, OCallApi>(params, ocall_api)?;
+		validator.set_state(validation_state);
+		info!("Found already initialized light client with Genesis Hash: {:?}", genesis);
+		info!("light client state: {:?}", validator);
+		Ok(validator)
+	} else {
+		init_validator::<B, OCallApi>(params, ocall_api)
 	}
 }
 
-fn init_validator<B: Block>(params: LightClientInitParams<B::Header>) -> Result<B::Header>
+fn init_validator<B, OCallApi>(
+	params: LightClientInitParams<B::Header>,
+	ocall_api: Arc<OCallApi>,
+) -> Result<LightValidation<B, OCallApi>>
 where
+	B: Block,
 	NumberFor<B>: finality_grandpa::BlockNumberOps,
+	OCallApi: EnclaveOnChainOCallApi,
 {
-	match params {
-		LightClientInitParams::Grandpa { authorities, genesis_header, authority_proof } => {
-			let mut validator = GrandpaLightValidation::<B>::new();
-
-			validator.initialize_relay(genesis_header, authorities, authority_proof)?;
-			LightClientSeal::<B, GrandpaLightValidation<B>>::seal_to_static_file(
-				validator.clone(),
-			)?;
-
-			return Ok(validator.latest_finalized_header(validator.num_relays()).unwrap())
+	// TODO: initialize relay will be more generic, so there will be changes here with issue #776
+	let validator: LightValidation<B, OCallApi> = match params {
+		LightClientInitParams::Grandpa { genesis_header, authorities, authority_proof } => {
+			let finality: Arc<Box<dyn Finality<B> + Sync + Send + 'static>> =
+				Arc::new(Box::new(GrandpaFinality {}));
+			let mut validator = LightValidation::<B, OCallApi>::new(ocall_api, finality);
+			validator.initialize_grandpa_relay(genesis_header, authorities, authority_proof)?;
+			validator
 		},
-	}
+		LightClientInitParams::Parachain { genesis_header } => {
+			let finality: Arc<Box<dyn Finality<B> + Sync + Send + 'static>> =
+				Arc::new(Box::new(GrandpaFinality {}));
+			Arc::new(Box::new(ParachainFinality {}));
```

Comment:
There seems to be a slight error here, we probably only want the second line (doing nothing atm), not the instantiation of the `GranpaFinality`? Maybe we should write a test, so that such an oversight is detected